### ISSUE TITLE
@ParcelConstructor doesn't work with Kotlin kapt

### ISSUE THREAD

**Issue by lukaspili** [johncarl81/parceler#143]
Hello,
Kapt, the annotation processor provided by Kotlin does not work with `@ParcelConstructor`.
I'm guessing parceler is matching constructor parameter name with the field name in the class.
Kapt doesn't respect the parameter names in the constructor, and it becomes "arg0", "arg1", "arg2".

I had the same issue with my own annotation processor, and I submitted a bug: https://youtrack.jetbrains.com/issue/KT-9609

However there is now way to say if that bug will be fixed for the 1.0 release or not.
In the meantime I'm wondering if you could implement a quick fix like the one I did on my project: https://github.com/lukaspili/Mortar-architect/blob/1.0_dev3.x/robot-compiler/src/main/java/architect/autostack/compiler/ConstructorParameterExtractor.java

Basically if we cannot match the parameter names with the fields names, we try to match the declaration order of parameters with the declaration order of fields.

Thanks.

**Comment by johncarl81** [comment]
That's unfortunate... Does `@ParcelConstructor` work when you annotate the constructor parameters with `@ParcelProperty`?

**Comment by lukaspili** [comment]
With `@ParcelProperty` the argX error goes away, but parceler and transfuse still crash.

```kotlin
@Parcel
data class Place @ParcelConstructor constructor(
        @ParcelProperty("id") val id: Int,
        @ParcelProperty("name") val name: String)
```

=> org.parceler.transfuse.util.TransfuseRuntimeException: Unable to perform code generation
=> warnings on all fields: Parceler: Reflection is required to access private field: int id, consider using non-private. (fields are public & final, and should be accessed by the constructor)


```kotlin
@Parcel
data class Place @ParcelConstructor constructor(
        @ParcelProperty("id") val id: Int,
        @ParcelProperty("name") val name: String,
        @ParcelProperty("images") val images: List<String>) // add List<String> field
```

=> error: Parceler: Unable to find read/write generator for type java.util.List<org.parceler.transfuse.adapter.ASTWildcardType@4759d11b> for List arg3


As for using the types rather than order to match arguments/fields with different names, I'm not sure that would be better. It's very common to have several Int / String fields, so the solution wouldn't work most of the times. I guess the best compromise is the current implementation with `@ParcelProperty` on constructor arguments.

**Comment by johncarl81** [comment]
Do you know how Kotlin handles generic types?  What does `List<String>` get converted into?

**Comment by lukaspili** [comment]
Ah yes kotlin actually converts to List<? extends String> (https://youtrack.jetbrains.com/issue/KT-5792).
Using MutableList rather than List solves the error.

The only thing left now is the parceler warning about reflection:
`Parceler: Reflection is required to access private field: int id, consider using non-private.`

Any ideas?
Thanks

**Comment by johncarl81** [comment]
I've seen success with people using the `BEAN` seralization strategy which relies on Kotlin's generated getter and setters instead of fields... does this work for you?:

```java
@Parcel(Serialization.BEAN)
data class Place @ParcelConstructor constructor(...)
```

**Comment by lukaspili** [comment]
Warning is still present with `BEAN`.
Besides the properties are final because I'm using `val` instead of `var`, so setters won't work.

**Comment by johncarl81** [comment]
FYI: http://stackoverflow.com/questions/33891814/use-of-parceler-with-kotlin-data-class-with-constructor-for-serialization/33892435#33892435

Would you be willing to share a sample project that I can familiarize myself with Kotlin and these issues?  I feel that there are some subtle and simple tweaks that could be made to make Kotlin support easier.

**Comment by lukaspili** [comment]
Ok, makes sense :) I misunderstood that Kotlin public property is actually a private field with getter/[setter]. It actually works now, I didn't set `BEAN` on all data classes.
I'll provide a kotlin parceler sample project in few hours.
Thanks for the help!

**Comment by johncarl81** [comment]
Fantastic, thanks @lukaspili.

**Comment by neworld** [comment]
With Kotlin `1.0.0-beta-4584` version fixed list type error:

> error: Parceler: Unable to find read/write generator for type java.util.Listorg.parceler.transfuse.adapter.ASTWildcardType@4759d11b for List arg2

**Comment by qq157755587** [comment]
@lukaspili Could you share your sample project?

**Comment by pepyakin** [comment]
In 1.0.0-beta-4584 I should explicitly designate param site: 

```kotlin
@Parcel(Serialization.BEAN)
data class PinVerification @ParcelConstructor constructor(
  @param:ParcelProperty("identificationResult") 
  val identificationResult: IdentificationResult
)
```

Otherwise, I get "Error:Parceler: No corresponding property found for constructor parameter arg0".

**Comment by ZakTaccardi** [comment]
would be nice if this error would at least tell us the class where the error is
